### ISSUE TITLE
fatal bug in cpplab branch which prevents it working on linux

### ISSUE THREAD

**Issue by sg-s** [sg-s/xolotl#83]
## symptoms:

- MATLAB freezes or crashes
- random hangs
- sometimes works, but doesn't work a second time

## possible causes

it may be a memory freeing (`free()`) issue. it looks like at some point, I am trying to reference something that is in protected memory, or something -- hard to say

## why does it work on macOS?

no idea. no errors whatsoever. 

one possibility is that I'm using LLVM on macOS, and gcc on linux

**Comment by sg-s** [comment]
need to check if the `cpp-linux-debug` branch works on linux. if it does, then the bug lies somewhere in the C++ files that have been changed b/w `cpp-linux-debug` and `master`

**Comment by alec-hoyland** [comment]
Tried this on my Manjaro Linux laptop. Consistently dies during integration with an error along the lines of:

`free(): invalid next size (normal)`

```matlab

% last cpplab commit: ecfa8853f317f664ccca66b968e678b935924ec8
% last xolotl commit: 52d2b4c30a74e9bf63d0c2d567095faeb2d10943


% conversion from Prinz to phi
vol = 1; % this can be anything, doesn't matter
f = 14.96; % uM/nA
tau_Ca = 200;
F = 96485; % Faraday constant in SI units
phi = (2*f*F*vol)/tau_Ca;

addpath('~/code/xolotl')
addpath('~/code/cpplab')

x = xolotl;
x.cleanup;
x.add('AB','compartment','V',-65,'Ca',0.02,'Cm',10,'A',0.0628,'vol',vol,'phi',phi,'Ca_out',3000,'Ca_in',0.05,'tau_Ca',tau_Ca);
x.AB.add('Leak','gbar',1.0,'E',-50);

x.transpile;
x.compile;

x
% is fine up until here, integrating causes it to crash
V = x.integrate;
```

I'm pretty useless at this sort of stuff, but maybe [valgrind](http://valgrind.org/) would help?

**Comment by sg-s** [comment]
well, that's good to know, because now we've isolated the bug to <100 lines of C++ code

**Comment by alec-hoyland** [comment]
With (old) C++ code, this runs fine on my Manjaro Linux laptop.

```matlab

% last cpplab commit: ecfa8853f317f664ccca66b968e678b935924ec8
% last xolotl commit: c975df680a76115ca26b86fdcad5e0452415c1d7


% conversion from Prinz to phi
vol = 1; % this can be anything, doesn't matter
f = 14.96; % uM/nA
tau_Ca = 200;
F = 96485; % Faraday constant in SI units
phi = (2*f*F*vol)/tau_Ca;

addpath('~/code/xolotl')
addpath('~/code/cpplab')

x = xolotl;
x.cleanup;
x.add('AB','compartment','V',-65,'Ca',0.02,'Cm',10,'A',0.0628,'vol',vol,'phi',phi,'Ca_out',3000,'Ca_in',0.05,'tau_Ca',tau_Ca);
x.AB.add('Leak','gbar',1.0,'E',-50);

x.transpile;
x.compile;

x
% is fine up until here, integrating causes it to crash
V = x.integrate;

```

**Comment by sg-s** [comment]
fixed in 8e83eeb3347e06eca9413b97c7ae44d3b5924487
